plumbing/util.go

```diff
@@ -459,36 +457,27 @@ func CloneRepository(remote types.GitRemote, inMem bool, basePath string) (rep *
 	return
 }
 
-// IsBinary detects if data is a binary value based on:
-// http://git.kernel.org/cgit/git/git.git/tree/xdiff-interface.c?id=HEAD#n198
+// IsBinary returns whether a file contains a NUL byte near the beginning of the file.
 func IsBinary(filePath string) (bool, error) {
 	r, err := os.Open(filePath)
 	if err != nil {
 		return false, err
 	}
 	defer r.Close()
 
-	reader := bufio.NewReader(r)
-	c := 0
-	for {
-		if c == SNIFFLEN {
-			break
-		}
-
-		b, err := reader.ReadByte()
-		if err == io.EOF {
-			break
-		}
-		if err != nil {
-			return false, err
+	buf := make([]byte, 8000)
+	n, err := io.ReadFull(r, buf)
+	if err != nil {
+		if errors.Is(err, io.EOF) || errors.Is(err, io.ErrUnexpectedEOF) {
+			err = nil
+		} else {
+			err = fmt.Errorf("check for binary: %w", err)
 		}
-
-		if b == byte(0) {
-			return true, nil
+	}
+	for _, b := range buf[:n] {
+		if b == 0 {
+			return true, err
 		}
-
-		c++
 	}
-
-	return false, nil
+	return false, err
```

Comment:
This function is just a little hard for me to follow because of how `err` can waterfall down in different ways, but considering how short it is it's not blocking at all and could just be a difference in taste.

Reply:
Thanks for pointing that out! I added a comment because the `io.Reader` "return both data and an error sometimes" behavior is definitely confusing and contrary to most other Go APIs.